src/buildstream/storage/directory.py

```diff
@@ -16,328 +17,499 @@
+    """Raised by Directory functions.
+
+    It is recommended to handle this error and raise a more descriptive
+    user facing :class:`.ElementError` or :class:`.SourceError` from this error.
+
+    If this is not handled, the BuildStream core will fail the current
+    task where the error occurs and present the user with the error.
     """
 
-    def __init__(self, message, reason=None):
+    def __init__(self, message: str, reason: str = None):
         super().__init__(message, domain=ErrorDomain.VIRTUAL_FS, reason=reason)
 
 
+class FileType(FastEnum):
+    """Depicts the type of a file"""
+
+    DIRECTORY: int = 1
+    """A directory"""
+
+    REGULAR_FILE: int = 2
+    """A regular file"""
+
+    SYMLINK: int = 3
+    """A symbolic link"""
+
+    def __str__(self):
+        # https://github.com/PyCQA/pylint/issues/2062
+        return self.name.lower().replace("_", " ")  # pylint: disable=no-member
+
+
+class FileStat:
+    """Depicts stats about a file
+
+    .. note::
+
+       This object can be compared with the equality operator, two :class:`.FileStat`
+       objects will be considered equivalent if they have the same :class:`.FileType`
+       and have equivalent attributes.
+    """
+
+    def __init__(
+        self, file_type: int, *, executable: bool = False, size: int = 0, mtime: float = BST_ARBITRARY_TIMESTAMP
+    ) -> None:
+
+        self.file_type: int = file_type
+        """The :class:`.FileType` of this file"""
+
+        self.executable: bool = executable
+        """Whether this file is executable"""
+
+        self.size: int = size
+        """The size of the file in bytes"""
+
+        self.mtime: float = mtime
+        """The modification time of the file"""
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, FileStat):
+            return NotImplemented
+
+        return (
+            self.file_type == other.file_type
+            and self.executable == other.file_type
+            and self.size == other.size
+            and self.mtime == other.mtime
+        )
+
+
 class Directory:
+    """The Directory object represents a directory in a filesystem hierarchy
+
+    .. note::
+
+       Directory objects can be iterated over. Iterating over a directory object
+       will yeild strings depicting the entries in the given directory, similar
+       to ``os.listdir()``.
+    """
+
     def __init__(self, external_directory=None):
         raise NotImplementedError()
 
-    def descend(self, *paths: str, create: bool = False, follow_symlinks: bool = False):
+    def __iter__(self) -> Iterator[str]:
+        raise NotImplementedError()
+
+    ###################################################################
+    #                           Public API                            #
+    ###################################################################
+
+    def descend(self, path: str, *, create: bool = False, follow_symlinks: bool = False) -> "Directory":
```

Comment:
As this also accepts .. (protected against escaping), the name may be a bit off. descend is also not common terminology when working with directories, as far as I know. Could it make sense to rename it to open_directory()? This would be consistent with open_file() and would be the Directory equivalent of POSIX open(..., O_DIRECTORY).

Reply:
Done, let's call it `open_directory()` instead.